akka-actor/src/main/scala/akka/io/TcpConnection.scala

```diff
@@ -389,6 +398,12 @@ private[io] abstract class TcpConnection(val tcp: TcpExt, val channel: SocketCha
           if (data eq remainingData) this
           else new PendingBufferWrite(commander, data, ack, buffer, tail) // copy with updated remainingData
 
+        } else if (data.nonEmpty && buffer.isReadOnly) {
+          val head = data.headByteString
+          val remaining = data.drop(head.length)
+
+          //this will always start with the direct buffer, or will change over to atmost one
+          PendingBufferWrite(commander, remaining, ack, tail).doWrite(info)
```

Comment:
hmm, so normally there was recursion to 'writeToChannel' on data.nonEmpty, which should only happen if the first chunk of data was written in full to the channel (any incomplete writes or end-of-data stopped the recursion).

The *data.nonEmpty && buffer.isReadOnly, new PBW.doWrite(...)* part is effectively doing the same thing (but with a new PBW. This does feel like abuse of the existing code though, and is murky. ill fix it up.